File: main_console.rb
require 'trollop'

require_relative "logging"
require_relative "game_logic"
require_relative "score_analyser"
require_relative "ai1_player"
require_relative "console_human_player"

class ConsoleGame

	# Create game & players
	def initialize(opts)
		@game = GameLogic.new
		@game.messages_to_console(true)
		@game.new_game(opts[:size], opts.handicap)
		@goban = @game.goban
		@players = []
		2.times do |color|
  		@players[color] = opts.ai>color ? Ai1Player.new(@goban,color) : ConsoleHumanPlayer.new(@goban,color)
		end
		@game.load_moves(opts.load) if opts.load
		# if no human is playing we create one to allow human interaction
		@spectator = opts.ai >= 2 ? ConsoleHumanPlayer.new(@goban,-1) : nil
    @scorer = ScoreAnalyser.new
	end

  # Show prisoner counts during the game  
  def show_prisoners
    prisoners = @game.prisoners?
    prisoners.size.times do |c|
      puts "#{prisoners[c]} #{Grid::COLOR_NAMES[c]} (#{Grid::COLOR_CHARS[c]}) are prisoners"
    end
    puts ""
  end
  
	def propose_console_end
    text = @scorer.compute_score(@goban, @game.komi, @game.who_resigned)
    puts @goban.scoring_grid.to_text { |c| Grid::COLOR_CHARS[c] }
  	text.each { |line| puts line }

	  # We ask human players; AI always accepts
	  @players.each do |player|
		  if player.is_human and !player.propose_score
			  # Ending refused, we will keep on playing
			  @game.accept_ending(false)
			  return
		  end
	  end
    @game.accept_ending(true)
	end

	def get_move_or_cmd
		if !@spectator or @num_autoplay > 0
		  @num_autoplay -= 1
		  return @players[@game.cur_color].get_move
		else
		  return @spectator.get_move(@game.cur_color)
		end
	end

	def play_move_or_cmd(move)
    if move.start_with?("cont")
      @num_autoplay = move.split(":")[1].to_i
      @num_autoplay = 1 if @num_autoplay == 0 # no arg is equivalent to continue:1
    elsif move.start_with?("pris")
      show_prisoners
    elsif move.start_with?("hist")
    	puts @game.history_string
    elsif move == "dbg" then
      @goban.debug_display
    elsif move == "help" then # this help is for console only
      puts "Move (e.g. b3) or pass, undo, resign, history, dbg, log:(level)=(1/0), load:(moves), continue:(times)"
      puts "Four letter abbreviations are accepted, e.g. \"hist\" is valid to mean \"history\""
    else
	    @game.play_one_move(move)
	  end
	end

	def play_game
		@num_autoplay = 0
		while ! @game.game_ended
		  if @game.game_ending
		    propose_console_end
		    next
		  end
		  move = get_move_or_cmd
		  play_move_or_cmd(move)
		end
		puts "Game ended."
		puts @game.history_string
	end

end

opts = Trollop::options do
  opt :size, "Goban size", :default => 9
  opt :ai, "How many AI players", :default => 2
  opt :handicap, "Number of handicap stones", :default => 0
  opt :load, "Game to load like e4,c3,d5", :type => :string
end
puts "Command line options received: #{opts}"

# Start the game
ConsoleGame.new(opts).play_game

File: js/MainConsole.js
//Translated from main_console.rb using babyruby2js
'use strict';

var Grid = require('./Grid');
var main = require('./main');
//require 'trollop';
var GameLogic = require('./GameLogic');
var ScoreAnalyser = require('./ScoreAnalyser');
var Ai1Player = require('./Ai1Player');
var ConsoleHumanPlayer = require('./ConsoleHumanPlayer');

/** @class Create game & players
 */
function ConsoleGame(opts) {
    this.game = new GameLogic();
    this.game.messagesToConsole(true);
    this.game.newGame(opts['size'], opts.handicap);
    this.goban = this.game.goban;
    this.players = [];
    for (var color = 1; color <= 2; color++) {
        this.players[color] = ( opts.ai() > color ? new Ai1Player(this.goban, color) : new ConsoleHumanPlayer(this.goban, color) );
    }
    if (opts.load()) {
        this.game.loadMoves(opts.load());
    }
    // if no human is playing we create one to allow human interaction
    this.spectator = ( opts.ai() >= 2 ? new ConsoleHumanPlayer(this.goban, -1) : null );
    this.scorer = new ScoreAnalyser();
}
module.exports = ConsoleGame;

// Show prisoner counts during the game  
ConsoleGame.prototype.showPrisoners = function () {
    var prisoners = this.game.prisoners();
    for (var c = 1; c <= prisoners.length; c++) {
        console.log(prisoners[c] + ' ' + Grid.COLOR_NAMES[c] + ' (' + Grid.COLOR_CHARS[c] + ') are prisoners');
    }
    console.log('');
};

ConsoleGame.prototype.proposeConsoleEnd = function () {
    var text = this.scorer.computeScore(this.goban, this.game.komi, this.game.whoResigned);
    console.log(this.goban.scoringGrid.toText(function (c) {
        return Grid.COLOR_CHARS[c];
    }));
    for (var line, line_array = text, line_ndx = 0; line=line_array[line_ndx], line_ndx < line_array.length; line_ndx++) {
        console.log(line);
    }
    // We ask human players; AI always accepts
    for (var player, player_array = this.players, player_ndx = 0; player=player_array[player_ndx], player_ndx < player_array.length; player_ndx++) {
        if (player.isHuman && !player.proposeScore()) {
            // Ending refused, we will keep on playing
            this.game.acceptEnding(false);
            return;
        }
    }
    return this.game.acceptEnding(true);
};

ConsoleGame.prototype.getMoveOrCmd = function () {
    if (!this.spectator || this.numAutoplay > 0) {
        this.numAutoplay -= 1;
        return this.players[this.game.curColor].getMove();
    } else {
        return this.spectator.getMove(this.game.curColor);
    }
};

ConsoleGame.prototype.playMoveOrCmd = function (move) {
    if (move.startWith('cont')) {
        this.numAutoplay = parseInt(move.split(':')[1]);
        if (this.numAutoplay === 0) { // no arg is equivalent to continue:1
            this.numAutoplay = 1;
        }
    } else if (move.startWith('pris')) {
        return this.showPrisoners();
    } else if (move.startWith('hist')) {
        console.log(this.game.historyString());
    } else if (move === 'dbg') {
        return this.goban.debugDisplay();
    } else if (move === 'help') { // this help is for console only
        console.log('Move (e.g. b3) or pass, undo, resign, history, dbg, log:(level)=(1/0), load:(moves), continue:(times)');
        console.log('Four letter abbreviations are accepted, e.g. "hist" is valid to mean "history"');
    } else {
        return this.game.playOneMove(move);
    }
};

ConsoleGame.prototype.playGame = function () {
    this.numAutoplay = 0;
    while (!this.game.gameEnded) {
        if (this.game.gameEnding) {
            this.proposeConsoleEnd();
            continue;
        }
        var move = this.getMoveOrCmd();
        this.playMoveOrCmd(move);
    }
    console.log('Game ended.');
    console.log(this.game.historyString());
};

opts = main.Trollop.options(function () {
    var opts;
    opt('size', 'Goban size', {'default':9});
    opt('ai', 'How many AI players', {'default':2});
    opt('handicap', 'Number of handicap stones', {'default':0});
    return opt('load', 'Game to load like e4,c3,d5', {'type':'string'});
});
console.log('Command line options received: ' + opts);
// Start the game
new ConsoleGame(opts).playGame();
// E02: unknown method ai()
// E02: unknown method load()
// E02: unknown method opt(...)
// W02: Unknown class supposed to be attached to main: Trollop
// E02: unknown method options(...)